// ant-design-vue/lib/locale-provider/ta_IN.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ta_IN = require('../vc-pagination/locale/ta_IN');

var _ta_IN2 = _interopRequireDefault(_ta_IN);

var _ta_IN3 = require('../date-picker/locale/ta_IN');

var _ta_IN4 = _interopRequireDefault(_ta_IN3);

var _ta_IN5 = require('../time-picker/locale/ta_IN');

var _ta_IN6 = _interopRequireDefault(_ta_IN5);

var _ta_IN7 = require('../calendar/locale/ta_IN');

var _ta_IN8 = _interopRequireDefault(_ta_IN7);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = {
  locale: 'ta',
  Pagination: _ta_IN2['default'],
  DatePicker: _ta_IN4['default'],
  TimePicker: _ta_IN6['default'],
  Calendar: _ta_IN8['default'],
  // locales for all comoponents
  global: {
    placeholder: 'தேதியைத் தேர்ந்தெடுக்கவும்'
  },
  Table: {
    filterTitle: 'பட்டியலை மூடு',
    filterConfirm: 'சரி',
    filterReset: 'மீட்டமை',
    emptyText: 'தகவல் இல்லை',
    selectAll: 'அனைத்தையும் தேர்வுசெய்',
    selectInvert: 'தலைகீழாக மாற்று',
    sortTitle: 'தலைப்பை வரிசைப்படுத்தவும்'
  },
  Modal: {
    okText: 'சரி',
    cancelText: 'ரத்து செய்யவும்',
    justOkText: 'பரவாயில்லை, சரி'
  },
  Popconfirm: {
    okText: 'சரி',
    cancelText: 'ரத்து செய்யவும்'
  },
  Transfer: {
    titles: ['', ''],
    notFoundContent: 'உள்ளடக்கம் கிடைக்கவில்லை',
    searchPlaceholder: 'இங்கு தேடவும்',
    itemUnit: 'தகவல்',
    itemsUnit: 'தகவல்கள்'
  },
  Upload: {
    uploading: 'பதிவேற்றுகிறது...',
    removeFile: 'கோப்பை அகற்று',
    uploadError: 'பதிவேற்றுவதில் பிழை',
    previewFile: 'கோப்பை முன்னோட்டமிடுங்கள்'
  },
  Empty: {
    description: 'தகவல் இல்லை'
  },
  Icon: {
    icon: 'உருவம்'
  },
  Text: {
    edit: 'திருத்து',
    copy: 'நகல் எடு',
    copied: 'நகல் எடுக்கப்பட்டது',
    expand: 'விரிவாக்கவும்'
  },
  PageHeader: {
    back: 'பின் செல்லவும்'
  }
};
